### PR TITLE
Obsolete `ISourceGenerator` APIs

### LINKED ISSUE #65106: Proposal: obsolete ISourceGenerator, SyntaxReceiver and all non-incremental APIs
## Background and Motivation

The non-incremental generator APIs, namely `ISourceGenerator` & friends, are fundamentally not scalable and suffer from pretty big performance issues, which makes it so that they should literally just never be used. Them existing poses several issues:
- There's plenty of docs and blog posts around showing how to use these APIs. Especially new developers see them, think they're correct, and then proceed to write their own generators like that. Which means that:
  - They waste time learning how to use a fundamentally broken API
  - They will waste time when (ideally) migrating, as their architecture will not have been properly structured for incrementality
  - If they publish their generators, users will use them, potentially in large solutions, VS will become slow, and they'll just open tickets for VS without realizing the issue is with one of the NuGet packages they're referencing
  - Even if they don't open tickets for VS, they'll still just enjoy a much worse developer UX while working
- There's literally nothing that clearly tells them that these types should just never be used anymore

This same scenario happened several times on C# Discord, and it keeps happening.
It would be nice to just finally obsolete these types 🙂

## Proposed API

```diff
namespace Microsoft.CodeAnalysis
{
+   [Obsolete("Non-incremental source generators should not be used, use IIncrementalGenerator instead.")]
    public interface ISourceGenerator
    {
    }

+   [Obsolete("Non-incremental source generators should not be used, use IIncrementalGenerator instead.")]
    public interface ISyntaxReceiver
    {
    }

+   [Obsolete("Non-incremental source generators should not be used, use IIncrementalGenerator instead.")]
    public delegate ISyntaxReceiver SyntaxReceiverCreator();

+   [Obsolete("Non-incremental source generators should not be used, use IIncrementalGenerator instead.")]
    public interface ISyntaxContextReceiver
    {
    }

+   [Obsolete("Non-incremental source generators should not be used, use IIncrementalGenerator instead.")]
    public delegate ISyntaxContextReceiver? SyntaxContextReceiverCreator();

+   [Obsolete("Non-incremental source generators should not be used, use IIncrementalGenerator instead.")]
    public readonly struct GeneratorInitializationContext
    {
    }

+   [Obsolete("Non-incremental source generators should not be used, use IIncrementalGenerator instead.")]
    public readonly struct GeneratorExecutionContext
    {
    }

+   [Obsolete("Non-incremental source generators should not be used, use IIncrementalGenerator instead.")]
    public readonly struct GeneratorSyntaxContext
    {
    }

+   [Obsolete("Non-incremental source generators should not be used, use IIncrementalGenerator instead.")]
    public readonly struct GeneratorPostInitializationContext
    {
    }
}
```

## Risks

No risks. This is not a binary nor a source breaking change. If people wanted to still use them they can just suppress the warning. It seems like just letting everyone use `ISourceGenerator` with no warnings whatsoever is a bigger risk on its own.

### PR DESCRIPTION
**Pull request by jjonescz**
Resolves #65106. I marked the APIs as obsolete with a custom `DiagnosticId` which is suppressed across Roslyn projects. Alternative would be to switch from `ISourceGenerator` to `IIncrementalGenerator` to represent all source generators, but that seems complex, so I'm trying out this simpler way.

### CODE REVIEW

File: src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs
Comment (jjonescz) [reply src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:18]: Unfortunately, it seems that this is an older version of `ObsoleteAttribute` which doesn't have that property. But it would be great if that could be used, I think it could be used to suppress the warning globally across Roslyn.

Comment (Youssef1313) [reply src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:18]: @jjonescz I think we can define our own `ObsoleteAttribute` as follows:

```csharp
namespace System
{
    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Struct | AttributeTargets.Enum |
        AttributeTargets.Interface | AttributeTargets.Constructor | AttributeTargets.Method | AttributeTargets.Property | AttributeTargets.Field | AttributeTargets.Event | AttributeTargets.Delegate,
        Inherited = false)]
    internal sealed class ObsoleteAttribute : Attribute
    {
        public ObsoleteAttribute()
        {
        }

        public ObsoleteAttribute(string? message)
        {
            Message = message;
        }

        public ObsoleteAttribute(string? message, bool error)
        {
            Message = message;
            IsError = error;
        }

        public string? Message { get; }

        public bool IsError { get; }

        public string? DiagnosticId { get; set; }

        public string? UrlFormat { get; set; }
    }
}
```

and apply it `[Obsolete(SourceGenerationMessages.UseIncrementalGenerator, DiagnosticId = "ID goes here")]`

Comment (jjonescz) [reply src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:18]: Thanks @Youssef1313, that looks good, I tried it. But declaring this type creates ambiguity with the original which is especially problematic in Visual Basic projects (which see this type due to `InternalsVisibleTo`) - they emit error BC30560 which cannot be suppressed.

Comment (Youssef1313) [reply src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:18]: One solution I can think of is declaring `ObsoleteAttribute` with `file` modifier. Though since we are obsoleting APIs in different files, the code will have to be repeated.

Comment (jjonescz) [reply src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:18]: Seems that such type is not picked up by the compiler - it doesn't generate obsolete warnings.

Comment (Youssef1313) [reply src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:18]: image

Strange. It's picked up for me in a new netcoreapp3.1 app

Comment (jjonescz) [reply src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:18]: Maybe because you are in one file?

Comment (jjonescz) [reply src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:18]: Oh, no, I had a bug in my `ObsoleteAttribute` definition (introduced while playing with it). It indeed works with `file` modifier. Thanks.

Comment (jjonescz) [reply src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs:18]: Except it seems it doesn't work across library boundaries.

File: src/Compilers/Core/Private/ObsoleteAttribute.cs
Comment (jjonescz) [inline src/Compilers/Core/Private/ObsoleteAttribute.cs:11]: Manually adding new definition of `ObsoleteAttribute` which has `DiagnosticId` so the warning can be disabled with one line in `.globalconfig`. That avoids lots of warning suppressions across the Roslyn repo (plus might be useful for customers).

File: src/Compilers/CSharp/Portable/Symbols/Symbol.cs
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Symbols/Symbol.cs:1264]: Needed to add suppressions around `<see cref="ObsoleteAttribute"/>` because `cref` considers even inaccessible types (the custom `ObsoleteAttribute` is not accessible in this project).

File: src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj
Comment (jjonescz) [inline src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj:3]: Unfortunately, I needed to create new project which contains the custom `internal class ObsoleteAttribute`, because if it was included in `Microsoft.CodeAnalysis`, it would be visible to all other IVT projects, leading to ambiguity warnings (or even errors in Visual Basic).

Comment (jcouv) [reply src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj:3]: It would be good to document (somehow) what is the purpose of this project. What types can/should be added here?

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 4)

File: src/Compilers/CSharp/Portable/Symbols/Symbol.cs
Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Symbols/Symbol.cs:1264]: > What happens in the resulting xml docs? Which symbol does it end up binding to?

It just refers to `System.ObsoleteAttribute`, I don't think it's possible to tell which one:

```xml
<member name="P:Microsoft.CodeAnalysis.CSharp.Symbol.ObsoleteState">
  <summary>
    True if this symbol has been marked with the
    <see cref="T:System.ObsoleteAttribute"/>
    attribute. This property returns
    <see cref="F:Microsoft.CodeAnalysis.ThreeState.Unknown"/>
    if the
    <see cref="T:System.ObsoleteAttribute"/>
    attribute hasn't been cracked yet.
  </summary>
</member>
```

> Is there a way we could update the cref to use unambiguous syntax instead?

I tried but haven't found a way.

File: src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj
Comment (jjonescz) [reply src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj:3]: It's shortly documented in `<PackageDescription>` field in this file - it's for private utility types, i.e., types that should not be IVT to other assemblies unlike types in `Microsoft.CodeAnalysis`. I'm not sure if I should be that specific in `<PackageDescription>`, I guess that's visible on NuGet. But I can add a comment.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 4)

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for second review. Thanks

File: src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj
Comment (jjonescz) [reply src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj:16]: I assumed the DLL needs to be packed and shipped since it's referenced by `Microsoft.CodeAnalysis.Common`?

Comment (Youssef1313) [reply src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj:15]: I'm not sure, but my thought was this will just be a small dll included inside Microsoft.CodeAnalysis.Common package.

@RikkiGibson thoughts?

Comment (jjonescz) [reply src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj:15]: Hm, didn't realize I could do that

Comment (RikkiGibson) [reply src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj]: Makes sense to me at a glance but I'm not super familiar with the plan to introduce a new CodeAnalysis dll. That's generally treated as a pretty big change especially if the compiler uses the dll.

Comment (jcouv) [reply src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj]: Good catch @RikkiGibson. I didn't think about that, but the additional dll would not fly with VS/RPS.
@jjonescz Is there an alternative design that doesn't involve a new dll?

Comment (jjonescz) [reply src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj]: > Is there an alternative design that doesn't involve a new dll?

Not this simple. Another alternative would be using `[Obsolete]` without `DiagnosticId` and suppress diagnostics via `#pragma`s manually everywhere (lots of places). Or replacing usages of `ISourceGenerator` in Roslyn by `IIncrementalGenerator` everywhere and wrapping `ISourceGenerator`s in an `IIncrementalGenerator` adapter (currently it's the other way around - `IIncrementalGenerator`s are wrapped as `ISourceGenerator`s) - not sure this is even possible, plus it might be a lot of work.

File: src/Compilers/CSharp/Portable/Symbols/Symbol.cs
Comment (sharwell) [inline src/Compilers/CSharp/Portable/Symbols/Symbol.cs:1259]: 📝 This can be resolved by placing a using alias directive at the top of the file. Both code and documentation comments prefer local aliases over other resolutions, and the alias will be resolved considering accessibility.

File: src/NuGet/Microsoft.Net.Compilers.Toolset/CoreClrCompilerArtifacts.targets
Comment (sharwell) [inline src/NuGet/Microsoft.Net.Compilers.Toolset/CoreClrCompilerArtifacts.targets:14]: @dotnet/roslyn-compiler this should be run through RPS before merging. I think it's going to trigger an assembly load regression, and I'm not sure the benefits of this change will warrant approving the new assembly. If it does get pre-approval, obviously no reason to block on this aspect.

Comment (Youssef1313) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/CoreClrCompilerArtifacts.targets:14]: @jjonescz I guess it will be less noisy to go back to the NuGet approach. I wasn't aware that adding a new dll is so tricky.

@sharwell does that work well?

Comment (sharwell) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/CoreClrCompilerArtifacts.targets:14]: I'm not sure what the NuGet approach is

Comment (sharwell) [reply src/NuGet/Microsoft.Net.Compilers.Toolset/CoreClrCompilerArtifacts.targets:14]: That wouldn't make any difference. The test looks at assemblies loaded in VS, regardless of source location.

File: src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj
Comment (cston) [reply src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj]: Could we use the BCL System.ObsoleteAttribute (to avoid a new DLL), but include `DiagnosticId` on .NET Core only, to allow easier suppression of warnings for .NET Core scenarios?
```csharp
    [Obsolete(Constants.UseIncrementalGenerator
#if NETCOREAPP
        , DiagnosticId = "ROSLYN0001"
#endif
        )]
    public readonly struct GeneratorExecutionContext { ... }
```

**Comment by sharwell** [comment]
@jjonescz Another option is to add ISourceGenerator to the list of APIs that are banned in analyzer projects:
https://github.com/dotnet/roslyn-analyzers/pull/6115

File: src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj
Comment (jjonescz) [reply src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj]: Yes, but I would still need to add lots of `#pragma warning disable` to build Roslyn itself. That's what I'm trying to avoid here. It's no problem for me to do that, I'm just not sure it would be approved.

Comment (jcouv) [reply src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj]: Maybe we could suppress the obsolete diagnostic for this API globally, instead of doing it in source at every location?

Comment (jjonescz) [reply src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj]: Without `DiagnosticId` set in `ObsoleteAttribute`, that could be done only by suppressing all obsolete warnings, I think. That's probably undesirable.

File: src/Compilers/Core/Portable/SourceGeneration/ISourceGenerator.cs
Comment (jasonmalinowski) [inline src/Compilers/Core/Portable/SourceGeneration/ISourceGenerator.cs:22]: It seems like this would be a fairly noisy block -- because ISourceGenerator is also used as the "currency" for anything running generators, including incremental generators, since we wrap them. As long as any _implementor_ of ISourceGenerator is getting flagged via any of the other obsoletions, then I'd remove this to avoid a bunch of unnecessary noise.

File: src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj
Comment (jasonmalinowski) [reply src/Compilers/Core/Private/Microsoft.CodeAnalysis.Private.csproj:3]: If we did go this approach this is still missing a bit more of a "why?" here -- we needed this because we had to redefine a framework type and that caused downstream problems. But for example it'd be safe to remove if we retargeted back to something that defined the DiagnosticId on the ObsoleteAttribute.

**Review by jasonmalinowski — COMMENTED** [review]
I'd be really curious what the alternative of just #pragma disabling the various uses of ISourceGenerator would look like in Roslyn (especially if we didn't obsolete the interface itself per my other comment.) Having an extra binary like this seems like a lot of work to avoid the one-time code change.

**Comment by jjonescz** [comment]
Thanks everyone for feedback. Let's try @sharwell's suggestion https://github.com/dotnet/roslyn/pull/66462#issuecomment-1421172603:
- Banned ISourceGenerator APIs: https://github.com/dotnet/roslyn-analyzers/pull/6478
- Extracted docs updates from this PR: https://github.com/dotnet/roslyn/pull/66752

**Pull request closed without merging**
